### ISSUE TITLE
`new` fails for subclasses

### ISSUE THREAD

**Issue by jph00** [pytorch/pytorch#47186]
## 🐛 Bug

`new` fails for subclasses

## To Reproduce

```python
import torch
class _T(torch.Tensor): ...
t = tensor([1.]).as_subclass(_T)
t.new([1,2])
```

This returns an empty tensor.

## Expected behavior

It should return

    tensor([1., 2.])

The behavior is correct when not using a subclass:

```python
t = tensor([1.])
t.new([1,2])
```

## Environment

PyTorch 1.7 on Ubuntu 20.04.

## Additional context

I remember we've discussed `new` with @soumith before - it's an important method for easily constructing new tensors with the same device etc as existing tensors. I think it ought to work the same on subclasses as on `Tensor`.

cc @hameerabbasi @rgommers

**Comment by hameerabbasi** [comment]
Should also be fixed by #47114

**Comment by jph00** [comment]
Thanks @hameerabbasi . Are you aware of any methods that are impacted by this? I'd like to add workarounds for them in `TensorBase` until the fix is in the release version.

**Comment by hameerabbasi** [comment]
Here's a list of the ones I fixed:

- `_is_view`
- `get_device`
- `has_names`
- `data_ptr`
- `dim`
- `__float__`
- `__complex__`
- `__int__`
- `__index__`
- `__invert__`
- `element_size`
- `item`
- `new`
- `new_tensor`
- `tolist`

**Comment by jph00** [comment]
Thanks that's very helpful.

**Comment by jph00** [comment]
BTW is `new` and `new_tensor` the same?

**Comment by hameerabbasi** [comment]
> BTW is `new` and `new_tensor` the same?

They seem to be almost the same, but I'd recommend `new_tensor` as it's documented.
